pkg/objstore/caching/readers.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package caching
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	tsdberrors "github.com/prometheus/prometheus/tsdb/errors"
+
+	"github.com/thanos-io/thanos/pkg/cache"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// Gives access to sequentially read blocks. Offsets are aligned to blockSize.
+type blocksProvider interface {
+	blockAt(offset int64) ([]byte, error)
+	close() error
+}
+
+// io.Reader implementation that uses blocks from blocksProvider.
+type blocksReader struct {
+	blockSize int64
+	blocks    blocksProvider
+
+	// Offset for next read, used to find correct block to return data from.
+	readOffset int64
+
+	// Remaining data to return from this reader. Once zero, this reader reports EOF.
+	remaining int64
+}
+
+func newBlocksReader(blockSize int64, blocks blocksProvider, readOffset, remaining int64) *blocksReader {
+	return &blocksReader{
+		blockSize:  blockSize,
+		blocks:     blocks,
+		readOffset: readOffset,
+		remaining:  remaining,
+	}
+}
+
+func (c *blocksReader) Close() error {
+	return c.blocks.close()
+}
+
+func (c *blocksReader) Read(p []byte) (n int, err error) {
+	if c.remaining <= 0 {
+		return 0, io.EOF
+	}
+
+	currentBlockOffset := (c.readOffset / c.blockSize) * c.blockSize
+	currentBlock, err := c.blocks.blockAt(currentBlockOffset)
+	if err != nil {
+		return 0, errors.Wrapf(err, "read position: %d", c.readOffset)
+	}
+
+	offsetInBlock := int(c.readOffset - currentBlockOffset)
+	toCopy := len(currentBlock) - offsetInBlock
+	if toCopy <= 0 {
+		// This can only happen if block's length is not blockSize, and reader is told to read more data.
+		return 0, errors.Errorf("no more data left in block at position %d, block length %d, reading position %d", currentBlockOffset, len(currentBlock), c.readOffset)
+	}
+
+	if len(p) < toCopy {
+		toCopy = len(p)
+	}
+	if c.remaining < int64(toCopy) {
+		toCopy = int(c.remaining) // Conversion is safe, c.remaining is small enough.
+	}
+
+	copy(p, currentBlock[offsetInBlock:offsetInBlock+toCopy])
+	c.readOffset += int64(toCopy)
+	c.remaining -= int64(toCopy)
+
+	return toCopy, nil
+}
+
+// Blocks provider using blocks cached in memory.
+type cachedBlocks struct {
+	// Mapping of blockSize-aligned offsets to keys in hits.
+	offsetsKeys  map[int64]string
+	blocks       map[string][]byte
+	bytesCounter prometheus.Counter
+}
+
+func newCachedBlocksProvider(offsetsKeys map[int64]string, blocks map[string][]byte, bytesCounter prometheus.Counter) *cachedBlocks {
+	return &cachedBlocks{
+		offsetsKeys:  offsetsKeys,
+		blocks:       blocks,
+		bytesCounter: bytesCounter,
+	}
+}
+
+func (c cachedBlocks) blockAt(offset int64) ([]byte, error) {
+	b := c.blocks[c.offsetsKeys[offset]]
+	if b == nil {
+		return nil, errors.Errorf("cachedBlocks: block for offset %d not found", offset)
+	}
+	c.bytesCounter.Add(float64(len(b)))
```

Comment:
This counter overcounts. The problem is that:
1. `blockAt()` can be potentially called multiple times for the same block (depends on the buffer passed to `Read()`)
2. If it's the first or last block, the reader may actually read less then whole bucket size

A different option may have `blockAt()` returning 3rd parameter to let the caller know whether the block has been picked from the cache or not. The caller can then correctly increment the counter when the block comes from the cache.

Reply:
Correct. This is fixed by simpler reader implementation.